### PR TITLE
Define the list of available eventHub events

### PR DESCRIPTION
**Pull request by RubenRBS**
Early stage PR to make `eventHub.on(event, callback)` properly typed.

There are some CI issues still, will fix in the next few days before merging

For now, the kind of errors that this finds are:
| Problem | Annotation |
|----- | ----- |
| Event missing from list | imagen |
| Mismatching parameter type | imagen |
| Arity mismatch | imagen |

### CODE REVIEW

File: static/events.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        compilerId: number,
+        lineNumber: number,
+        colBegin: number,
+        colEnd: number,
+        reveal: boolean,
+        sender: string
+    ) => void;
+    ppViewClosed: (compilerId: number) => void;
+    ppViewOpened: (compilerId: number) => void;
+    // TODO: See if this is necessary. Nothing emits this event
```
Comment (jeremy-rifkin) [inline static/events.ts:115]: I do believe this is emitted: https://github.com/compiler-explorer/compiler-explorer/blob/335f683b6b065072061496dd63e31201bece108d/static/panes/pp-view.ts#L88-L96

File: static/events.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    ppViewOptionsUpdated: (compilerId: number, options: string, recompile: boolean) => void;
+    requestCompilation: (editorId: number, treeId: boolean | number) => void;
+    requestMotd: () => void;
+    requestSettings: () => void;
+    resendCompilation: (compilerId: number) => void;
+    requestCompiler: (compilerId: number) => void;
+    // TODO: Bug? There are no listeners for this event, only Pane.ts emits it
```
Comment (jeremy-rifkin) [inline static/events.ts:122]: Listened here https://github.com/compiler-explorer/compiler-explorer/blob/335f683b6b065072061496dd63e31201bece108d/static/panes/compiler.js#L2231-L2239

File: static/events.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    toolInputViewClosed: (compilerId: number, toolId: number, input: string) => void;
+    toolInputViewCloseRequest: (compilerId: number, toolId: number) => void;
+    // TODO: Only emitted
+    toolInputViewOpened: (toolId: number) => void;
+    toolOpened: (compilerId: number, toolState: unknown) => void;
+    toolSettingsChange: (compilerId: number) => void;
+    treeClose: (treeId: number) => void;
+    treeCompilerEditorExcludeChange: (treeId: number, compilerId: number, editorId: number) => void;
+    treeCompilerEditorIncludeChange: (treeId: number, compilerId: number, editorId: number) => void;
+    // Todo: Only emitted
```
Comment (jeremy-rifkin) [inline static/events.ts:149]: Listened here https://github.com/compiler-explorer/compiler-explorer/blob/335f683b6b065072061496dd63e31201bece108d/static/hub.ts#L157-L163

**Review by jeremy-rifkin — COMMENTED** [review]
This looks amazing, this is really exciting!

File: static/events.ts (hunk 3 above)
Comment (RubenRBS) [reply static/events.ts:149]: Ohh. I know why I couldn't find any of this events. I grepped in single-line mode

Comment (RubenRBS) [reply static/events.ts:149]: This means I might have missed some events! But hopefuly they show up as errors if not present

File: static/panes/ast-view.ts [hunk 4]
```diff
@@ -227,7 +227,7 @@ export class Ast extends MonacoPane<monaco.editor.IStandaloneCodeEditor, AstStat
                     code.source.to.line < code.source.from.line + 100
                 ) {
                     for (let i = code.source.from.line; i <= code.source.to.line; ++i) {
-                        if (colours[i - 1] !== undefined) {
+                        if (!(i - 1 in colours)) {
```
Comment (RubenRBS) [inline static/panes/ast-view.ts:230]: This is the only change I'm iffy about as it's more than just type changes, but eslint was complaining

File: types/features/filters.interfaces.ts [hunk 5]
```diff
@@ -1,13 +1,40 @@
...
-    directives: boolean;
     labels: boolean;
-    optOutput: boolean;
     libraryCode: boolean;
+    directives: boolean;
+    commentOnly: boolean;
     trim: boolean;
+}
+
+export type ParseFilters = CompilerFilters & {
```
Comment (RubenRBS) [inline types/features/filters.interfaces.ts:37]: Separated this as the keys of this one do not show up in `static/`

File: static/panes/ast-view.ts (hunk 4 above)
Comment (jeremy-rifkin) [reply static/panes/ast-view.ts:230]: I think it'd be just `i - 1 in colours` without the negation

Comment (RubenRBS) [reply static/panes/ast-view.ts:230]: Oops, you're right

**Comment by RubenRBS** [comment]
Weird. `make check` ran `ts-compile` and did not output anything, but now there are annotations!

File: static/panes/ast-view.ts (hunk 4 above)
Comment (RubenRBS) [inline static/panes/ast-view.ts:230]:
```suggestion
                        if (i - 1 in colours) {
```

File: static/event-hub.ts [hunk 6]
```diff
@@ -26,7 +26,7 @@ import GoldenLayout from 'golden-layout';
 import * as Sentry from '@sentry/browser';
 
 import {Hub} from './hub';
-import {Events} from './events';
+import {EventsList} from './events-list';
```
Comment (matsjla) [inline static/event-hub.ts:29]: Maybe EventMap would be a better name here? Events have a name->data mapping which makes Map/Record/etc more suitable in my opinion

Comment (RubenRBS) [reply static/event-hub.ts:29]: Sure!

**Comment by RubenRBS** [comment]
Finally ready to get merged

**Comment by partouf** [comment]
will test later

**Pull request merged**